File: OrangeJuice.Server.Core.FSharp/Cache/MemoryCacheClient.fs
namespace OrangeJuice.Server.FSharp.Cache

open System
open System.Runtime.Caching

open OrangeJuice.Server.Cache

type MemoryCacheClient(cache : ObjectCache) =
    interface ICacheClient with
        member this.AddOrGetExisting<'T>(key : string, valueFactory : Func<'T>) : 'T =
           (this :> ICacheClient).AddOrGetExisting(key, valueFactory, fun () -> new CacheItemPolicy())
        
        member this.AddOrGetExisting<'T>(key : string, valueFactory : Func<'T>, policyFactory : Func<CacheItemPolicy>) : 'T =
            let newValue = new Lazy<'T>(valueFactory)
            let policy = policyFactory.Invoke()
            let value = cache.AddOrGetExisting(key, newValue, policy) :?> Lazy<'T>

            let (|??) l r = if l <> null then l else r
            (value |?? newValue).Value

File: OrangeJuice.Server.Core/Cache/MemoryCacheClient.cs
using System;
using System.Runtime.Caching;

namespace OrangeJuice.Server.Cache
{
	public sealed class MemoryCacheClient : ICacheClient
	{
		private readonly ObjectCache _cache;

		public MemoryCacheClient(ObjectCache cache)
		{
			_cache = cache;
		}

		public T AddOrGetExisting<T>(string key, Func<T> valueFactory)
		{
			return AddOrGetExisting(key, valueFactory, () => new CacheItemPolicy());
		}

		public T AddOrGetExisting<T>(string key, Func<T> valueFactory, Func<CacheItemPolicy> policyFactory)
		{
			var newValue = new Lazy<T>(valueFactory);
			var policy = policyFactory();
			var value = (Lazy<T>)_cache.AddOrGetExisting(key, newValue, policy);
			return (value ?? newValue).Value;
		}
	}
}